src/components/ChatContent/Messages/BlockMenu/ReportDialog.vue

```diff
@@ -0,0 +1,300 @@
+      <a
+        target="_blank"
+        href="http://www.eternagame.org/web/player/36921/"
+      >Hoglahoo</a>
+      and
+      <a target="_blank" href="http://www.eternagame.org/web/player/48290/">LFP6</a>.
+    </p>
+    <p>User: {{userToReport.username}}</p>
+    <div style="text-align: center; margin-top: 15px;">
+      <div style="display: inline-block; margin: 0px 12px;">
+        <input type="checkbox" v-model="ignore">
+        <label for="report-ignore">Ignore User</label>
+      </div>
+      <div style="display: inline-block; margin: 0px 12px;">
+        <input type="checkbox" v-model="report">
+        <label for="report-report">Report User</label>
+      </div>
+    </div>
+    <div style="margin-top: 15px;" v-show="report">
+      <label for="report-message">Additional Comments:</label>
+      <textarea v-model="reportComments" style="resize: none; height: 60px;"></textarea>
+    </div>
+    <div style="margin-top: 15px;">
+      <div
+        class="blue-button blue-button-interactive"
+        style="width: 40%; display: inline-block; vertical-align: middle;"
+        @click="closeModal"
+      >Cancel</div>
+      <div
+        class="green-button green-button-interactive"
+        style="width: 40%; display: inline-block; float: right; vertical-align: middle;"
+        @click="submit"
+      >Continue</div>
+    </div>
+  </modal>
+</template>
+
+<script lang="ts">
+  import { Component, Prop } from 'vue-property-decorator';
+  import Vue from '@/types/vue';
+  import { Client } from 'irc-framework';
+  import parseUsername from '@/tools/parseUsername';
+  import User from '../../../../types/user';
+  import Message from '../../../../types/message';
+
+@Component({})
+  export default class ReportDialog extends Vue {
+  report = false;
+
+  ignore = false;
+
+  message: Message | null = null;
+
+  get userToReport(): User {
+    return this.message ? this.message.user : User.annonymous;
+  }
+
+  $refs!: {
+    reportModal: HTMLFormElement;
+  };
+
+  reportComments: string = '';
+
+  submit() {
+    if (this.ignore) {
+      this.$store.commit('ignore', this.userToReport.username);
+    }
+    const client = this.$store.state.client!;
+    if (this.report) {
+      client.say(
+        '#ops-notifications',
+        `[REPORT] Reporting ${this.userToReport.username} (${
+          this.userToReport.uid
+        }) by ${this.$store.state.currentUser.username} (${
+          this.$store.state.currentUser.uid
+        }).\r\n`,
+      );
+      if (this.message) {
+        client.say(
+          '#ops-notifications',
+          `[REPORTED MESSAGE] ${this.message!.message}\r\n`,
+        );
+      }
+      client.say(
+        '#ops-notifications',
+        `[REPORT REASON] ${this.reportComments}\r\n`,
+      );
+    }
+    this.closeModal();
+  }
+
+  closeModal() {
+    this.$modal.hide('reportModal');
+    this.reportComments = '';
+  }
+
+  created() {
+    this.$store.subscribe((mutation, state) => {
```

Comment:
I ended up creating an `open` function to which you pass the data, as opening the modal feels in nature more like an action than a state.

Reply:
_When in Rome, do as the Romans do_

How reactive frameworks like Vue are designed such that you have a state, and you declare your view. The view is automatically synced with the state, so all you need to do is change the state and the view knows how to keep itself updated. "Opening the modal" is an action, yes - however you should think about this a bit differently. Instead, consider it as changing "is the modal showing". As much as possible, you should rely on state mutations and declarative UI. Otherwise you get back to the realm of jQuery code where you have bits of code everywhere modifying UI who knows where and you don't know where anything comes from.

Reply:
Another reason for keeping the open function is that the props are only mutated once, all at the same time, when we open the modal. Then, when we mutate them from inside the modal, we sync them back outside so that they don't break, but we don't actually use the new value outside of the component.
I think I'll move to other stuff for now, and return to it fresh to rethink it.

Reply:
I think that if I pass the `ignore` and `report` properties in an object (which feels more fitting now that I think about it), I will be able to modify them from within the component. I'll try that.

Reply:
It seems that it references the store in `submit` in order to dispatch the report. Should that be an event we listen to in `ChatContent`, or would that be overkill?
I think I prefer to keep it inside the component, but I'm not sure about it.

Reply:
> I think that if I pass the `ignore` and `report` properties in an object (which feels more fitting now that I think about it), I will be able to modify them from within the component. I'll try that.
I still have the problem of the visibility, which is changed from within the component. I could raise an event that it was closed, but I just don't think it's worth it; Opening and closing the modal is implemented multiple events (opening and closing), and we pass a property that only describes one of them, but still has to be up to date with the modal's status no matter which of them happened.
I think that the correct approach would still be to pass the message as a property, just like I pass data that doesn't have anything with the opening process to the modal itself, but have the visibility set by a function.
The problem then is what to do with the defaults (ignore & report); I think that they should go into the function, as they get disconnected from the property (at least logically) after the dialog opens and the user is allowed to change them.

Reply:
> It seems that it references the store in submit in order to dispatch the report. Should that be an event we listen to in ChatContent, or would that be overkill?
Help me out - what exactly is the event you'd be listening for, and what would you do when receiving it? I'm pretty sure you're right that it should be within the component but I'm not sure

> I still have the problem of the visibility, which is changed from within the component. I could raise an event that it was closed, but I just don't think it's worth it;
Shouldn't .sync handle that? Otherwise please clarify what the issue is.

> The problem then is what to do with the defaults (ignore & report); I think that they should go into the function, as they get disconnected from the property (at least logically) after the dialog opens and the user is allowed to change them
I think you're thinking too far into this. Let me put it in perspective.
```vue
<template>
  <dialog :open.sync="open" :report.sync="report" :ignore.sync="ignore">
</template>
<script>
class Foo {
    open = false;
    report = false;
    ignore = false;
}
</script>
```
Even though you don't _use_ the values after you set them, they still represent the state of the UI.

I mean it's personal opinion, but I just really feel like it's cleaner to just do everything declaratively. Even if it's a bit verbose, it's uniform and clear where things are coming from. Instead of "hey, there's a function, where does it get called?" it's "ok there's a property for it, I just look to see whats bound to it, and any changes to that state change it".